### DISCUSSION TITLE
Support flow duplication (RFC)

### DISCUSSION THREAD

**Discussion by david-zacharias** [directus/directus#18559] [Feature Requests]
### Summary

Allow users to duplicate flows via Directus Studio UI

### Motivation

Users may want to try out changes to a flow but want to keep the original one as a backup.

### UI

In alignment to duplication of fields, the duplication can be triggered via the Select-Menu of each Flow in the List of Flows:

<img width="250" alt="Bildschirmfoto 2023-05-11 um 15 44 58" src="https://github.com/directus/directus/assets/1903179/e5337c95-cf29-4efb-9fb4-5927e98e1032">

After clicking a modal should be opened, asking the user to name the new flow. As default the flows current name appended with `(Copy)` is prefilled.

Users must confirm which leads to flow duplication or cancel.

### UX

The new flow should be disabled by default, to not have same flows being active in parallel.

### Logic

A flow consits of one `flow` record and 0 to n `operation` records, wich are linked via relations:

Relational properties of a **Flows**:

| Property  | Note |
| ---- | ---- |
| `operation` | UUID of the operation connected to the trigger in the flow. |

Relational properties an **Operation**:

| Property | Contains |
| ---- | ---- |
|`resolve` | The operation triggered when the current operation succeeds (or then logic of a condition [operation](https://docs.directus.io/reference/system/operations.html#the-operation-object)). Primary key of an operation. |
|`reject` | The operation triggered when the current operation fails (or otherwise logic of a condition [operation](https://docs.directus.io/reference/system/operations.html#the-operation-object)). Primary key of an operation. |
| `flow` | The flow containing this operation. Many-to-one to [flows](https://docs.directus.io/reference/system/flows.html#the-flow-object). |

The current implementation of the Directus Studio UI for managing flows uses a PATCH request on `flows/<the-flow-uuid>` with nested data like:

```json
{
  "name": "Some flow",
  "operation": { 
    "id": "cdd92780-2c55-4a52-bb1a-9f80a3d55900",
    "flow": "1c797550-ff17-49ff-93e8-2dbe4f3b8931",
    "resolve": {
      "id": "8fb91151-07b7-4e09-9e04-bbdb86f85ea7"
      "flow": "1c797550-ff17-49ff-93e8-2dbe4f3b8931"
      "reject": null
    }
  }
}
```

This allows to change / add the flow and all operations in one transaction.

Data from the backend (`flows/<uuid>?fields=*,operations.*`) is received as flat list of operations with references (*). We must transpose this list to a a tree like structure with `flow.operation` as root node then further spanned via `operation.resolve` and `operation.reject`. From all nodes (including the `flow` itself) the `id` property must be excluded and `flow`-property from each operation. 

To disable the duplicated flow by default `flow.status` should be set to `inactive`.

This new structure must be posted to `/flows` which then recursively creates all operations including new uuids for each operation. 

#### Edge case
There are may dangling operations which are configured but not linked to another operation or the flow (`flow.operation`) itself. These will not be included in the structured data described before, and must be added separately. Maybe via `flow.operations`-relational-property

(*) We could make use `?field`-Query  to receive nested relational field data, but this is probably less performant. And we must know / compute the maximum depth of the operations tree in advance.

### Must Have:
- A Flow including all operations must be copied

### Should Have:
- C Possibility to give the copied flow a new name before advance of duplication

### Could Have:
- D

### Won't Have:
- E

### Drawbacks

I see no drawbacks, the feature is something users presumably feel natural to exist and can use without additional learning.

### Alternatives

No alternatives considered 

### Adoption Strategy

Neither data model nor existing business logic is altered; no breaking changes.

**Comment by rijkvanzanten** [comment]
Question @david-zacharias! Should we consider this as a way to duplicate _anything anywhere_? A flow with operations from the app's perspective is really no difference from any item with a related item. If we treat this is adding support for duplication _in general_ and then enabling it on the directus_flows system table, I think we can unlock way more value 🙂

**Reply by david-zacharias** [reply]
Interesting idea and great high-level view 👀  So to say we add a deep copy functionality for items?

My first thought was that there is a difference between system collections (= modelled by the team; relational fields are known in advance / _at coding time_) and user collections. For the latter case we'd need to first retrieve the schema to identify relational fields which must be excluded / overwritten before writing the new item. 

Second thought: Users can already extend any system collection with custom fields including relational ones which leads to the need for these "generically" build transformations anyway - at least if there isn't a completely different approach for how item duplication could work (Ideas?).

This said, yes, the implementation should be applicable to other (all) items 🤓 

I guess `directus_fields` can be used to determin all relational fields. Problem: We do not want to copy **all** relations; for example, a deep copy would also copy the user referenced via `user_created`. How can we determinitically determin which relation to keep and which to clone 🤔 Could `item_duplication_fields` on the collection be used?

**Reply by rijkvanzanten** [reply]
> My first thought was that there is a difference between system collections (= modelled by the team; relational fields are known in advance / at coding time) and user collections. For the latter case we'd need to first retrieve the schema to identify relational fields which must be excluded / overwritten before writing the new item.

Yeah! We currently already have that config for the "Save as Copy" functionality, so to me this duplicate piece would effectively just be that "Save as Copy" without having any other changes that you applied. The only thing I'm not sure about is how to handle unique constraints 🤔 With the save-as-copy flow, you have a chance to manually change a handful of fields, but the duplication might fail if you have a unique constraint

Although on second thought, that's what the duplication fields setting is for! You could exclude the unique field from being duplicated to prevent that problem 🤔

**Reply by david-zacharias** [reply]
>We currently already have that config for the "Save as Copy" functionality,

Do you think the UI/UX for duplicating flows should be the same as "Save and Copy" for items? Personally, I'd prefer duplication via the selection menu as implemented for fields.

>Although on second thought, that's what the duplication fields setting is for! You could exclude the unique field from being duplicated to prevent that problem.

Agree, it is currently also possible to misconfigure the _duplication fields_, so "Save as Copy" would fail - maybe another area to improve: Excluding / disabling fields with unique constraint from the duplication fields_ selection list at all 🤓 

For me the difference between the existing "Save as Copy" and the Duplication was  that "Save and Copy" does not copy any field by default, one must configure `item_duplication_fields`. Duplication would copy all fields – however, leading to the problems we discussed above.

So, let's say we configure (via migration) default values for `item_duplication_fields` on `directus_flows` and `directus_operations` collections. These fields will then be included in the duplication. We still have to build the tree structure of `flows` and all belonging `operations`. So for a general approach duplication (in contrast to "Save and Copy") would recursively resolve all relational fields which are listed in `item_duplication_fields` of the respective collection.

Currently the UI for selecting _Item Duplication Fields_ is disabled for system collections. To allow users to have custom fields being included in the copy, the UI (and api) should allow inclusion of custom fields.

A duplication button with the above functionality could be added to any collection, so it's versatile. Nevertheless, I'd see adding a duplication button for custom collections in scope of a follow up ticket, because we may have to think about UX a bit more. Users probably expect a duplication button to duplicate things **without prior configuration** - and as said above for custom collections no field is copied by default.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by rijkvanzanten** [comment]
Lets keep the discussion going in the original post so we don't end up with two threads with two upvote numbers 🙂 See https://github.com/directus/directus/discussions/14405
